### PR TITLE
Centralized components

### PR DESCRIPTION
**Pull request by jonathan2222**
Made it easier to create components.

### CODE REVIEW

File: LambdaEngine/Include/ECS/ComponentArray.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	public:
+		ComponentArray();
+		virtual ~ComponentArray() = default;
+
+		Comp& Insert(Entity entity, Comp comp);
+		void Remove(Entity entity);
+		
+		Comp& GetData(Entity entity);
+
+		void EntityDestroyed(Entity);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentArray.h]:
```suggestion
		void EntityDestroyed(Entity) override final;
```

File: LambdaEngine/Include/ECS/ComponentManager.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include "ComponentArray.h"
+#include "Component.h"
+#include "Defines.h"
+
+namespace LambdaEngine
+{
+	class ComponentManager
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentManager.h]: How about renaming it `ComponentStorage`?

File: LambdaEngine/Include/ECS/ComponentManager.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	public:
+		DECL_UNIQUE_CLASS(ComponentManager);
+		ComponentManager();
+		~ComponentManager();
+
+		template<typename Comp>
+		void RegisterComponentType();
+
+		template<typename Comp>
+		Comp& AddComponent(Entity entity, Comp component);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentManager.h]: Have you considered making components const references in the 'add' functions?

File: LambdaEngine/Include/ECS/Entity.h [hunk 4]
```diff
@@ -5,4 +5,5 @@
 namespace LambdaEngine
 {
 	typedef uint32 Entity;
+	typedef uint64 Index;
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/Entity.h]: I don't think this adds any clarity. Instead of renaming the type, you could rename variables, eg. `uint64 componentIdx`.

File: LambdaEngine/Include/ECS/ECSCore.h [hunk 5]
```diff
@@ -26,7 +27,18 @@ namespace LambdaEngine
...
+        // Add a component to a specific entity. 
+        template<typename Comp>
+        Comp& AddComponent(Entity entity, Comp component);
+
+        // Remove a component from a specific entity.
+        template<typename Comp>
+        bool RemoveComponent(Entity entity);
+
+        // Remove a specific entity.
+        void RemoveEntity(Entity entity);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ECSCore.h:41]: Why is this needed, and why should it coexist with `EnqueueEntityDeletion`?

File: LambdaEngine/Include/ECS/ComponentManager.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		ComponentArray<Comp>* compArray = dynamic_cast<ComponentArray<Comp>*>(m_ComponentArrays[index]);
+
+		return compArray->GetData(entity);
+	}
+
+	template<typename Comp>
+	inline bool ComponentManager::HasType()
+	{
+		std::type_index id = Comp::s_TID;
+		return m_CompTypeToArrayMap.find(id) != m_CompTypeToArrayMap.end();
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentManager.h]:
```suggestion
		return m_CompTypeToArrayMap.find(Comp::s_TID) != m_CompTypeToArrayMap.end();
```

I think the same thing can be done above, in `ComponentManager::GetComponent`.

File: LambdaEngine/Include/ECS/ComponentManager.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+	template<typename Comp>
+	inline Comp& ComponentManager::GetComponent(Entity entity)
+	{
+		std::type_index id = Comp::s_TID;
+		VALIDATE_MSG(m_CompTypeToArrayMap.find(id) != m_CompTypeToArrayMap.end(), "Trying to fetch a component which was not registered!");
+
+		// Fetch the corresponding ComponentArray for that component type.
+		uint32 index = m_CompTypeToArrayMap[id];
+		ComponentArray<Comp>* compArray = dynamic_cast<ComponentArray<Comp>*>(m_ComponentArrays[index]);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentManager.h]:
```suggestion
		ComponentArray<Comp>* compArray = static_cast<ComponentArray<Comp>*>(m_ComponentArrays[index]);
```

`dynamic_cast` does runtime checks to ensure that the cast is safe, we don't need that, since we know the cast is safe.

File: LambdaEngine/Include/ECS/ComponentManager.h [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+		void RegisterComponentType();
+
+		template<typename Comp>
+		Comp& AddComponent(Entity entity, Comp component);
+
+		template<typename Comp>
+		void RemoveComponent(Entity entity);
+
+		template<typename Comp>
+		Comp& GetComponent(Entity entity);
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentManager.h]: As mentioned verbally, I think we should have an `GetArray` function. This would allow systems to cache the lookup of the component array.

File: LambdaEngine/Source/ECS/ComponentManager.cpp [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#include "ECS/ComponentManager.h"
+
+using namespace LambdaEngine;
+
+LambdaEngine::ComponentManager::ComponentManager()
```
Comment (TheoBerlin) [inline LambdaEngine/Source/ECS/ComponentManager.cpp]: Declare the constructor as default instead (in the header file, do: `ComponentManager() = default;`)

File: LambdaEngine/Include/ECS/ECSCore.h [hunk 10]
```diff
@@ -77,4 +90,29 @@ namespace LambdaEngine
...
+        ComponentAdded(entity, Comp::s_TID);
+
+        return comp;
+    }
+
+    template<typename Comp>
+    inline bool ECSCore::RemoveComponent(Entity entity)
+    {
+
+        if (m_ComponentManager.HasType<Comp>())
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ECSCore.h]:
```suggestion
        if (m_ComponentManager.HasType<Comp>())
```

Remove empty line.

File: LambdaEngine/Include/ECS/ComponentArray.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+		Entity lastEntity = m_IndexToEntity[lastIndex];
+		m_EntityToIndex[lastEntity] = currentIndex;
+		m_IndexToEntity[currentIndex] = lastEntity;
+		
+		// Remove the deleted component's entry.
+		m_EntityToIndex.erase(entity);
+		m_IndexToEntity.erase(lastIndex);
+
+		// Decrease size counter.
+		m_Size--;
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentArray.h]: You can remove this and in line 68 do: `Index lastIndex = --m_Size;`

File: LambdaEngine/Include/ECS/ComponentArray.h [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+		// Resize if more data is added.
+		if (m_Size+1 > m_Data.GetSize())
+			m_Data.Resize(m_Size + m_Capacity);
+
+		// Get new index and add the component to that position.
+		Index newIndex = m_Size;
+		m_EntityToIndex[entity] = newIndex;
+		m_IndexToEntity[newIndex] = entity;
+		memcpy(&m_Data[newIndex], &comp, sizeof(Comp));
+		m_Size++;
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentArray.h]: You can remove this and in line 53 do: `Index newIndex = m_Size++;`

File: CrazyCanvas/Source/CrazyCanvas.cpp [hunk 13]
```diff
@@ -46,6 +46,10 @@
 #include <rapidjson/rapidjson.h>
 #include <rapidjson/writer.h>
 
+// TEMP!!
```
Comment (TheoBerlin) [inline CrazyCanvas/Source/CrazyCanvas.cpp]: I'm not a big fan of pushing temporary test code to master.

**Review by TheoBerlin — CHANGES_REQUESTED** [review]

File: LambdaEngine/Include/ECS/ComponentManager.h (hunk 3 above)
Comment (jonathan2222) [reply LambdaEngine/Include/ECS/ComponentManager.h]: Good point, done.

File: LambdaEngine/Include/ECS/ECSCore.h (hunk 5 above)
Comment (jonathan2222) [reply LambdaEngine/Include/ECS/ECSCore.h:41]: I think it is more clear to have those functions. First of all, the name is more clear than EnqueueEntityDeletion and it is good to be able to remove specific components and not the whole entity.
And the components within the ComponentManager needs to be removed.
I would like to replace the EnqueueEntityDeletion but I was told to not remove any code and instead only add the things I needed.

File: LambdaEngine/Include/ECS/ComponentManager.h [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+		void EntityDeleted(Entity entity);
+
+		template<typename Comp>
+		bool HasType();
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ComponentManager.h]: Can be const.

File: LambdaEngine/Include/ECS/ECSCore.h [hunk 15]
```diff
@@ -38,7 +50,6 @@ namespace LambdaEngine
         void PerformRegistrations();
 
         // Enqueues an entity deletion
```
Comment (TheoBerlin) [inline LambdaEngine/Include/ECS/ECSCore.h]: Move or remove comment.

**Review by TheoBerlin — APPROVED** [review]
